### PR TITLE
Additional analysis fixes

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Fix C26110,C28023,C28193,C6001,C6031,C6054

Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

**Comment by dthaler** [comment]
@Alan-Jowett merged 312, please rebase and resolve conflicts.  Thanks!

File: libs/platform/user/ebpf_extension_user.c [hunk 1]
```diff
@@ -217,12 +217,15 @@ ebpf_provider_load(
 
 Done:
     ebpf_lock_unlock(&_ebpf_provider_table_lock, state);
-    ebpf_free(local_extension_provider);
+    // Wrap in conditional to resolve C6101.
```
Comment (dthaler) [inline libs/platform/user/ebpf_extension_user.c:220]: https://docs.microsoft.com/en-us/cpp/code-quality/c6101?view=msvc-160  says C6101 means "A successful path through the function does not set the named _Out_ parameter." which seems unrelated to 222.  Is C6101 the correct number and if so do you understand what it was complaining about?

Comment (Alan-Jowett) [reply libs/platform/user/ebpf_extension_user.c:220]: I think what is happening is analysis sees the local_extension_provider as an alias for the out parameter and incorrectly sees the ebpf_free as making it invalid. Wrapping it makes analysis see the out as still valid.

**Comment by Alan-Jowett** [comment]
Looks like the recent rebase broke stuff. Will take a closer look when I get home.

File: libs/platform/user/ebpf_extension_user.c (hunk 1 above)
Comment (Alan-Jowett) [reply libs/platform/user/ebpf_extension_user.c]: Here's the full error text:
E:\ebpf-for-windows\libs\platform\user\ebpf_extension_user.c(159): warning C6101: Returning uninitialized memory '*local_extension_provider'.  A successful path through the function does not set the named _Out_ parameter. 

Note:
1) local_extension_provider is not an out parameter.
2) When the free occurs on the success path, local_extension_provider == NULL.

Comment (Alan-Jowett) [reply libs/platform/user/ebpf_extension_user.c:220]: image

Here is the detailed error message from the analysis output.

File: libs/platform/ebpf_serialize.c [hunk 2]
```diff
@@ -214,7 +216,10 @@ ebpf_deserialize_map_information_array(
     out_map_info = NULL;
 
 Exit:
-    ebpf_map_information_array_free(map_count, out_map_info);
+    // Wrap in conditional to resolve C6101.
+    if (result != EBPF_SUCCESS) {
```
Comment (dthaler) [inline libs/platform/ebpf_serialize.c:220]: what is the issue here?   the free function below should be annotated to accept null, and out_map_info is always initialized (line 139)

File: libs/platform/unit/platform_unit_test.cpp [hunk 3]
```diff
@@ -276,42 +276,63 @@ TEST_CASE("program_type_info_stored", "[platform]")
     ebpf_free(bind_program_information);
 }
 
+struct ebpf_security_descriptor_t_free
+{
+    void
+    operator()(ebpf_security_descriptor_t* p)
```
Comment (dthaler) [inline libs/platform/unit/platform_unit_test.cpp:282]: add SAL annotation to parameter p

File: libs/platform/unit/platform_unit_test.cpp [hunk 4]
```diff
@@ -276,42 +276,63 @@ TEST_CASE("program_type_info_stored", "[platform]")
...
+    REQUIRE(ebpf_validate_security_descriptor(sd_ptr.get(), sd_size) == EBPF_SUCCESS);
 
-    REQUIRE((result = ebpf_access_check(sd, 1, &generic_mapping), LocalFree(sd), result == EBPF_ACCESS_DENIED));
+    REQUIRE(ebpf_access_check(sd_ptr.get(), 1, &generic_mapping) == EBPF_ACCESS_DENIED);
 }
 
+struct ebpf_memory_descriptor_t_free
+{
+    void
+    operator()(ebpf_memory_descriptor_t* p)
```
Comment (dthaler) [inline libs/platform/unit/platform_unit_test.cpp:321]: add SAL annotation to parameter p

File: libs/platform/ebpf_serialize.c (hunk 2 above)
Comment (Alan-Jowett) [reply libs/platform/ebpf_serialize.c:220]: The issue is that code analysis gets confused by the _Post_ptr_invalid_ annotation and marks the _Out_ map_info parameter as uninitialized. 

Two mitigations are possible (until we get a fix/explanation from the code analysis team):
1) We can remove the post invalid qualification on all functions that free structures.
2) We can wrap calls to the free so they are only invoked on unsuccessful paths.

Comment (dthaler) [reply libs/platform/ebpf_serialize.c:220]: My opinion is that we should not make changes that affect the actual binaries, to work around code analysis flaws.  Wrapping calls to free affects the actual binary by increasing code complexity and size.

Comment (dthaler) [reply libs/platform/ebpf_serialize.c:220]: Is there any _Analysis_assume_ usable here?

Comment (shankarseal) [reply libs/platform/ebpf_serialize.c:220]: can you put analysis_assume that ouT_map_info != NULL if result is failure?

**Review by dthaler — APPROVED** [review]
